hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/utils/AzcopyToolHelper.java

```diff
@@ -0,0 +1,274 @@
+    if (!currentDir.isDirectory() && !currentDir.getName().equals(HADOOP_AZURE_DIR)) {
+      hadoopAzureDir = findHadoopAzureDir(currentDir);
+      if (hadoopAzureDir == null) {
+        throw new FileNotFoundException(HADOOP_AZURE_DIR + DIR_NOT_FOUND_ERROR);
+      }
+    } else {
+      hadoopAzureDir = currentDir;
+    }
+    return hadoopAzureDir;
+  }
+
+  private File findHadoopAzureDir(File dir) {
+    if (dir == null) {
+      return null;
+    }
+    File[] files = dir.listFiles();
+    if (files == null) {
+      return null;
+    }
+    for (File file : files) {
+      if (file.isDirectory() && file.getName().equals(HADOOP_AZURE_DIR)) {
+        return file;
+      } else {
+        File hadoopAzureDir = findHadoopAzureDir(file);
+        if (hadoopAzureDir != null) {
+          return hadoopAzureDir;
+        }
+      }
+    }
+    return null;
+  }
+
+  // Synchronized method to download azcopy executable if not present.
+  private void downloadAzcopyExecutableIfNotPresent()
+      throws IOException, InterruptedException {
+    // Check if azcopy directory is present in the hadoop-azure directory.
+    // If not present, create a directory and stop other process here only.
+    File azcopyDir = new File(azcopyDirPath);
+    if (!azcopyDir.exists()) {
+      azcopyDir.mkdir();
+    } else {
+      return;
+    }
+
+    // Check if azcopy tool is present in the azcopy directory.
+    File azcopyFile = new File(azcopyExecutablePath);
+    if (!azcopyFile.exists()) {
+      // If azcopy tool is not present, download and extract it
+      String[] downloadCmdArr = {AZCOPY_CMD_SHELL,
+          AZCOPY_CMD_OPTION, AZCOPY_DOWNLOAD_CMD};
+      Process downloadProcess = Runtime.getRuntime().exec(downloadCmdArr);
+      downloadProcess.waitFor();
+
+      // Extract the azcopy executable from the tarball
+      String extractCmd = EXTRACT_CMD + hadoopAzureDir.getAbsolutePath();
+      String[] extractCmdArr = {AZCOPY_CMD_SHELL,
+          AZCOPY_CMD_OPTION, extractCmd};
+      Process extractProcess = Runtime.getRuntime().exec(extractCmdArr);
+      extractProcess.waitFor();
+
+      // Rename the azcopy_linux_amd64_* directory to 'azcopy' and move it to the hadoop-azure directory
+      String renameCmd = MOVE_CMD + hadoopAzureDir.getAbsolutePath() + AZCOPY_DOWNLOADED_DIR_NAME + azcopyDirPath;
+      String[] renameCmdArr = {AZCOPY_CMD_SHELL, AZCOPY_CMD_OPTION, renameCmd};
+      Process renameProcess = Runtime.getRuntime().exec(renameCmdArr);
+      renameProcess.waitFor();
+
+      // Remove the downloaded tarball and azcopy folder
+      String cleanupCmd = REMOVE_CMD + hadoopAzureDir.getAbsolutePath() + AZCOPY_DOWNLOADED_TAR_NAME;
+      String[] cleanupCmdArr = {AZCOPY_CMD_SHELL, AZCOPY_CMD_OPTION, cleanupCmd};
+      Process cleanupProcess = Runtime.getRuntime().exec(cleanupCmdArr);
+      cleanupProcess.waitFor();
+
+      // Set the execute permission on the azcopy executable
```

Comment:
if (!azcopyFile.exists()) {
        executeCommand(AZCOPY_CMD_SHELL, AZCOPY_CMD_OPTION, AZCOPY_DOWNLOAD_CMD);
        executeCommand(AZCOPY_CMD_SHELL, AZCOPY_CMD_OPTION, EXTRACT_CMD + hadoopAzureDir.getAbsolutePath());
        executeCommand(AZCOPY_CMD_SHELL, AZCOPY_CMD_OPTION, MOVE_CMD + hadoopAzureDir.getAbsolutePath() + AZCOPY_DOWNLOADED_DIR_NAME + azcopyDirPath);
        executeCommand(AZCOPY_CMD_SHELL, AZCOPY_CMD_OPTION, REMOVE_CMD + hadoopAzureDir.getAbsolutePath() + AZCOPY_DOWNLOADED_TAR_NAME);
        executeCommand(AZCOPY_CMD_SHELL, AZCOPY_CMD_OPTION, CHMOD_CMD + azcopyDirPath);
    }
}

private void executeCommand(String... command) throws IOException, InterruptedException {
    Process process = new ProcessBuilder(command).start();
    process.waitFor();  duplication of code can be removed using this simplification

Reply:
Nice suggestion, Taken